metric/emitter/prometheus.go

```diff
@@ -0,0 +1,188 @@
+package emitter
+
+import (
+	"fmt"
+	"net/http"
+
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/atc/db"
+	"github.com/concourse/atc/metric"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type PrometheusEmitter struct {
+	buildsStarted     prometheus.Counter
+	buildsFinished    prometheus.Counter
+	buildsSucceeded   prometheus.Counter
+	buildsErrored     prometheus.Counter
+	buildsFailed      prometheus.Counter
+	buildsAborted     prometheus.Counter
+	buildsFinishedVec *prometheus.CounterVec
+	buildDurationsVec *prometheus.SummaryVec
+}
+
+type PrometheusConfig struct {
+	Address string `long:"prometheus-address" description:"Address to listen on to expose Prometheus metrics."`
+}
+
+func init() {
+	metric.RegisterEmitter(&PrometheusConfig{})
+}
+
+func (config *PrometheusConfig) Description() string { return "Prometheus" }
+func (config *PrometheusConfig) IsConfigured() bool  { return config.Address != "" }
+
+func (config *PrometheusConfig) NewEmitter() (metric.Emitter, error) {
+	buildsStarted := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "started_total",
+		Help:      "Total number of Concourse builds started.",
+	})
+	prometheus.MustRegister(buildsStarted)
+
+	buildsFinished := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "finished_total",
+		Help:      "Total number of Concourse builds finished.",
+	})
+	prometheus.MustRegister(buildsFinished)
+
+	buildsSucceeded := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "succeeded_total",
+		Help:      "Total number of Concourse builds succeeded.",
+	})
+	prometheus.MustRegister(buildsSucceeded)
+
+	buildsErrored := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "errored_total",
+		Help:      "Total number of Concourse builds errored.",
+	})
+	prometheus.MustRegister(buildsErrored)
+
+	buildsFailed := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "failed_total",
+		Help:      "Total number of Concourse builds failed.",
+	})
+	prometheus.MustRegister(buildsFailed)
+
+	buildsAborted := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "concourse",
+		Subsystem: "builds",
+		Name:      "aborted_total",
+		Help:      "Total number of Concourse builds aborted.",
+	})
+	prometheus.MustRegister(buildsAborted)
+
+	buildsFinishedVec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "concourse",
+			Subsystem: "builds",
+			Name:      "finished",
+			Help:      "Count of builds finished across various dimensions.",
+		},
+		[]string{"team", "pipeline", "status"},
+	)
+	prometheus.MustRegister(buildsFinishedVec)
+	buildDurationsVec := prometheus.NewSummaryVec(
```

Comment:
I should probably make this a Histogram instead of a Summary (https://prometheus.io/docs/practices/histograms/)